data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,593 @@
+
+These two definitions are the same. The former explicitly binds the anonymous function to a name. The latter uses a more compact syntax and avoids the `fun` keyword and the arrow symbol.
+
+## Defining Local Functions
+
+A function may be defined locally. Just like any local definition, the function is only available inside its attached expression. Although local functions are often defined inside the function's scope, this is not a requirement. Here is a local function inside a toplevel expression:
+```ocaml
+# let sq x = x * x in sq 7 * sq 7;;
+- : int = 2401
+
+# sq;;
+Error: Unbound value sq
+```
+
+Calling `sq;;` gets an error because it was only defined locally.
+
+## Closures
+
+The following example illustrates a [*closure*](https://en.wikipedia.org/wiki/Closure_(computer_programming)) using [Same-Level Shadowing](#same-level-shadowing]:
+```ocaml
+# let d = 2 * 3;;
+val d : int = 6
+
+# let f x = x * d;;
+val f : int -> int = <fun>
+
+# f 7;;
+- : int = 42
+
+# let d = 10;;
+val d : int = 10
+
+# f 7;; (* What is the result? *)
+- : int = 42
+```
+
+Here is how this makes sense:
+1. Constant `d` is defined, and its value is 6.
+1. Function `f` is defined. It takes a single parameter `x` and returns its product by `d`.
+1. Compute `f` of 7, and its value is 42
+1. Create a new definition `d`, shadowing the first one
+1. Compute `f` of 7 again, the result is the same: 42
+
+Although the new definition of `d` *shadows* the first one, the original remains the one the function `f` uses. The `f` function's environment captures the first value of `d`, so every time you apply `f` (even after the second definition of `d`), you can be confident the function will behave the same. A closure is a pair containing the function code and an environment.
+
+However, all future expressions will use the new value of `d` (`10`), as shown here:
+
+```
+# let g = d * 3;;
+val g : int = 30
+```
+
+Partially applying parameters to a function creates a new closure. The environment is updated, but the function is unchanged.
+```ocaml
+# let max_42 = max 42;;
+val max_42 : int -> int = <fun>
+```
+
+Inside the `max_42` function, the environment contains an additional binding between the first parameter of `max` and the value 42.
+
+## Recursive Functions
+
+In order to perform iterated computations, a function may call itself. Such a function is called _recursive_. In OCaml, recursive functions must be defined and explicitly declared by using `let rec`. It is not possible to accidentally create recursion loops between functions. As a consequence, recursive functions can't be anonymous.
+
+The classic (and very inefficient) way to present recursion is using the function that computes the [Fibonacci Sequence](https://en.wikipedia.org/wiki/Fibonacci_sequence).
+```ocaml
+# let rec fibo n = if n <= 1 then n else fibo (n - 1) + fibo (n - 2);;
+val fibo : int -> int = <fun>
+
+# List.init 10 Fun.id |> List.map fibo;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
```

Comment:
At this point, `List.init` isn't covered in the Lists tutorial. `Fun.id` also isn't covered, as far as I can find.
If we keep this example this needs to be explained.

Reply:
I've given explaining it a shot. If we want to keep the `List.map fibo` part after the `pipe` (to illustrate the `pipe` in use again), then we'll need to explain why. If it's run without out, the result is the same, so why use it?

If  `List.map fibo;;` is run on it's own separately, it returns: `- : int list -> int list = <fun>`